decavision/dataset_preparation/generate_pseudolabels.py

```diff
@@ -0,0 +1,159 @@
+
+            Returns:
+                pseudo_data_path: A folder with both labeled and pseudo labeled images.
+        """
+        # Copy the training/labeled data to the destination folder where
+        # we will also store the pseudo labels.
+        copy_tree(train_data_path, pseudo_data_path)
+
+        # Load model
+        model = load_model(model_path, compile=False)
+        print("Loaded model.")
+
+        # Make dictionary for classes and their index
+        class_names = sorted(os.listdir(train_data_path))
+        class_dict = {}
+        index = 0
+        for cat in class_names:
+            class_dict[cat] = index
+            index += 1
+
+        print("Generating pseudo labels...")
+        # Generate pseudo labels
+        unlabeled_image_paths = os.listdir(unlabeled_path)
+        print("There are {} unlabeled images.".format(
+            len(unlabeled_image_paths)))
+        raw_predictions_paths = []
+        raw_predictions = []  # single confidence value of predicted class
+        raw_predictions_all = []  # confidences for all classes
+        for path in tqdm(unlabeled_image_paths[:100]):
+            # Load the image
+            img = np.array(self._load_img(os.path.join(unlabeled_path, path),
+                                          target_size=(299, 299)))
+            # Make predictions using trained model
+            y_pred = model.predict(np.expand_dims(img, axis=0), verbose=0)
+            # Get class index
+            y = np.argmax(y_pred)
+            # Get probability score
+            proba = y_pred[0][y]
+
+            raw_predictions.append(proba)
+            raw_predictions_all.append(y_pred)
+            raw_predictions_paths.append(os.path.join(unlabeled_path, path))
+
+            # The results variable should be the same as the class category
+            for class_name, index in class_dict.items():
+                if threshold:
+                    # For thresholding predictions
+                    if index == y and proba >= threshold:
+                        shutil.copy(os.path.join(unlabeled_path, path),
+                                    os.path.join(pseudo_data_path, class_name))
+                else:
+                    # For hard predictions
+                    if index == y:
+                        shutil.copy(os.path.join(unlabeled_path, path),
+                                    os.path.join(pseudo_data_path, class_name))
+        print("Generated pseudo labels.")
+
+        if not os.path.exists("outputs"):
+            os.mkdir("outputs")
+        if save_confidences_csv:
+            print("Saving CSV with pseudo predictions.")
+            data = {'filepaths': raw_predictions_paths,
+                    'class preds highest': raw_predictions,
+                    'class preds all': raw_predictions_all}
+            df = pd.DataFrame(data)
+            df.to_csv("outputs/pseudo_predictions.csv", index=False)
+
+        if plot_confidences:
```

Comment:
I would put the code to make the plots into a separate method. There are a few reasons why. First you seem to be repeating a lot of lines with the two types of plots so having a method to use would avoid that. Second people could plot the data as they like by calling the method instead of having to recalculate the pseudolabels every time. Finally this method is pretty big so splitting the logically distinct parts is a good thing to do.

Reply:
Done.